Implement the ability for user-defined type casting (#941)
yiisoft/db: docs/guide/en/schema/typecasting.md

use Yiisoft\Db\Connection\ConnectionInterface;

/** @var ConnectionInterface $db */
$command = $db->createCommand();
$command->insert('customer', [
    'name' => 'John Doe',
    'is_active' => true,
]);
$command->execute();
```

In the example above, the value of `is_active` is a `boolean`, but the column `is_active` can be of type `bit`.
The `CommandInterface::insert()` method will automatically cast the value to the correct type.

## Casting values retrieved from the database

When you retrieve a value from the database, the value can be returned in a different type than you expect.
For example, a value that is stored as an `float` in the database can be returned as a `string`. This is because 
the database driver do not convert some data types when retrieves values.

To ensure that the value is returned in the correct type, you can use `ColumnInterface::phpTypecast()` method to cast 
the value to the correct type.

```php
use Yiisoft\Db\Connection\ConnectionInterface;

/** @var ConnectionInterface $db */
$command = $db->createCommand('SELECT * FROM {{customer}} WHERE id = 1');

$row = $command->queryOne();
$isActive = $row['is_active'];

// Cast the value to the correct type

> $isActive = $db->getTableSchema('customer')->getColumn('is_active')->phpTypecast($isActive);

Reviewer: Is there a way to cast the whole row at once?

Author: No, but there is a task about how to do it automatically
https://github.com/yiisoft/db/issues/771

use Yiisoft\Db\Expression\Expression;
use Yiisoft\Db\Expression\ExpressionInterface;
use Yiisoft\Db\Schema\AbstractColumnSchema;

class PointColumn extends AbstractColumnSchema
{
    /**
     * @var string The default column abstract type
     */
    protected const DEFAULT_TYPE = 'point';

    /**
    * @param ExpressionInterface|Point|string|null $value
    */
    public function dbTypecast(mixed $value): ExpressionInterface|string|null
    {
        if ($value instanceof Point) {
            return new Expression('(:x,:y)', ['x' => $value->getX(), 'y' => $value->getY()]);
        }
    
        return $value;
    }

    /**
    * @param string|null  $value
    */
    public function phpTypecast(mixed $value): Point|null
    {
        if (is_string($value)) {
            [$x, $y] = explode(',', substr($value, 1, -1));

            return new Point((float) $x, (float) $y);
        }
    
        return $value;
    }
}

class Point
{
    public function __construct(
        private float $x,
        private float $y,
    ) {
    }

    public function getX(): float
    {
        return $this->x;
    }

    public function getY(): float
    {
        return $this->y;
    }
}
```

> Then it needs to use the custom column class in the database connection configuration.

Reviewer: ```suggestion
Then use the custom column class in the database connection configuration.
```

```php
use Yiisoft\Db\Constant\ColumnType;
use Yiisoft\Db\Pgsql\Column\ColumnFactory;
use Yiisoft\Db\Pgsql\Connection;
use Yiisoft\Db\Schema\Column\ArrayLazyColumn;
use Yiisoft\Db\Schema\Column\StructuredLazyColumn;
use Yiisoft\Db\Schema\Data\JsonLazyArray;

$columnFactory = new ColumnFactory(
    'columnClassMap' => [
        ColumnType::ARRAY => ArrayLazyColumn::class, // converts values to `LazyArray` objects 
        ColumnType::JSON => JsonLazyColumn::class, // converts values to `JsonLazyArray` objects
        ColumnType::STRUCTURED => StructuredLazyColumn::class, // converts values to `StructuredLazyArray` objects
    ],
);

// Create a database connection with the custom column factory
$db = new Connection($pdoDriver, $schemaCache, $columnFactory);

/** @var JsonLazyArray $tags `tags` column is of database type `json` */
$tags = $db->getTableSchema('customer')->getColumn('tags')->phpTypecast($row['tags']);

foreach ($tags as $tag) {
    echo $tag;
}
```

## Structured data types

Some databases support structured data types, such as `composite` types in Postgres. To cast a structured data type to

> a custom class, it needs to create a column class which extends `AbstractStructuredColumn` and override 

Reviewer: ```suggestion
a custom class, you need to create a column class which extends `AbstractStructuredColumn` and override 
```

                'currency_money' => new CurrencyMoney(...$value->getValue()),
                default => $value,
            };
        }
        
        return $value;
    }
}

class CurrencyMoney implements \JsonSerializable, \IteratorAggregate
{
    public function __construct(
        private float $value,
        private string $currencyCode = 'USD',
    ) {
    }

    public function getCurrencyCode(): string
    {
        return $this->currencyCode;
    }

    public function getValue(): float
    {
        return $this->value;
    }

    /**
     * `JsonSerializable` interface is implemented to convert the object to a database representation.
     */
    public function jsonSerialize(): array
    {
        return [
            'value' => $this->value,
            'currency_code' => $this->currencyCode,
        ];
    }
    
    /** 
     * Alternatively, `IteratorAggregate` interface can be implemented to convert the object to a database representation.
     */
    public function getIterator(): ArrayIterator
    {
        return new ArrayIterator([
            'value' => $this->value,
            'currency_code' => $this->currencyCode,
        ]);
    } 
}
```

> Then it needs to use the column class in the database connection configuration.

Reviewer: ```suggestion
Then use the column class in the database connection configuration.
```

```php
use Yiisoft\Db\Expression\Expression;
use Yiisoft\Db\Expression\ExpressionInterface;
use Yiisoft\Db\Pgsql\Data\StructuredParser;

class MyStructuredColumn extends AbstractStructuredColumn
{
    /**
    * @param string|null $value
    */
    public function phpTypecast(mixed $value): CurrencyMoney|StructuredLazyArray|null
    {
        if (is_string($value)) {
            $value = new StructuredLazyArray($value, $this->getColumns());
        
            return match ($this->getDbType()) {
                'currency_money' => new CurrencyMoney(...$value->getValue()),
                default => $value,
            };
        }
        
        return $value;
    }
}

class CurrencyMoney implements \JsonSerializable, \IteratorAggregate
{
    public function __construct(
        private float $value,
        private string $currencyCode = 'USD',
    ) {
    }

    public function getCurrencyCode(): string
    {
        return $this->currencyCode;
    }

    public function getValue(): float
    {
        return $this->value;
    }

    /**
     * `JsonSerializable` interface is implemented to convert the object to a database representation.
     */

>     public function jsonSerialize(): array

Reviewer 2: May be to add another example with `dbTypecast()` . This allow to keep DTO clear, without infrastructure things.

Author: This is how to convert object to array, then the array is typecasting and converting to string representation.

Also possible way to declare the properties as public.

https://github.com/yiisoft/db/blob/adf61039ab6dce56cfc08a2c1c3818df387297b5/src/Helper/DbArrayHelper.php#L235-L251
